File: dart/lib/otp_crypto/otp_cipher.dart
/// OTP Crypto – AES-256-CBC with PKCS#7 padding (via `encrypt` package)
/// --------------------------------------------------------------------
/// Thin wrapper around `package:encrypt` to perform AES-256-CBC encryption
/// and decryption with PKCS#7 padding. This class **does not** derive keys
/// or IVs; it only consumes the `encKey` (32B) and `iv` (16B) provided by
/// higher layers (HKDF + IV derivation).
///
/// SECURITY NOTES:
/// - Enforce key length (32) and IV length (16) before calling AES.
/// - This layer performs **no** MAC verification; always verify the HMAC tag
///   (Encrypt-then-MAC) *before* decryption at a higher layer.
/// - Catch and wrap low-level errors to avoid leaking internals.
///
/// HINTS:
/// - Use `encrypt(...)` to obtain ciphertext bytes for the wire header `c`.
/// - Use `decrypt(...)` only **after** tag verification passes.

import 'dart:typed_data';

import 'package:encrypt/encrypt.dart' as enc;

import 'errors.dart';

class OtpCipher {
  OtpCipher._(); // static-only

  /// Encrypts [plaintext] using AES-256-CBC with PKCS#7 padding.
  ///
  /// [encKey] must be exactly 32 bytes.
  /// [iv]     must be exactly 16 bytes.
  ///
  /// RETURNS: ciphertext bytes.
  ///
  /// Throws [InternalCryptoException] if encryption fails unexpectedly.
  static Uint8List encrypt({
    required Uint8List encKey,
    required Uint8List iv,
    required Uint8List plaintext,
  }) {
    _requireKeyIv(encKey, iv);

    try {
      final key = enc.Key(encKey);
      final ivObj = enc.IV(iv);

      // Explicitly request CBC + PKCS7 padding.
      final aes = enc.AES(
        key,
        mode: enc.AESMode.cbc,
        padding: 'PKCS7',
      );
      final encrypter = enc.Encrypter(aes);

      final encrypted = encrypter.encryptBytes(plaintext, iv: ivObj);
      return Uint8List.fromList(encrypted.bytes);
    } catch (e, st) {
      // Avoid surfacing library-specific errors.
      throw InternalCryptoException(cause: e, stackTrace: st);
    }
  }

  /// Decrypts [ciphertext] using AES-256-CBC with PKCS#7 padding.
  ///
  /// [encKey] must be exactly 32 bytes.
  /// [iv]     must be exactly 16 bytes.
  ///
  /// RETURNS: plaintext bytes.
  ///
  /// Throws [DecryptionFailedException] on any failure (including bad padding).
  /// IMPORTANT: Callers must have already verified the HMAC tag.
  static Uint8List decrypt({
    required Uint8List encKey,
    required Uint8List iv,
    required Uint8List ciphertext,
  }) {
    _requireKeyIv(encKey, iv);

    try {
      final key = enc.Key(encKey);
      final ivObj = enc.IV(iv);

      final aes = enc.AES(
        key,
        mode: enc.AESMode.cbc,
        padding: 'PKCS7',
      );
      final encrypter = enc.Encrypter(aes);

      // The `encrypt` package accepts raw bytes via `Encrypted`.
      final decrypted = encrypter.decryptBytes(enc.Encrypted(ciphertext), iv: ivObj);
      return Uint8List.fromList(decrypted);
    } catch (e, st) {
      // On any error, present a generic decryption failure.
      throw DecryptionFailedException(cause: e, stackTrace: st);
    }
  }

  // -- helpers ---------------------------------------------------------------

  /// Validates key and IV lengths for AES-256-CBC.
  ///
  /// HINT: Keep input validation close to crypto calls to fail fast.
  static void _requireKeyIv(Uint8List encKey, Uint8List iv) {
    if (encKey.length != 32) {
      throw ArgumentError.value(encKey.length, 'encKey.length', 'AES-256 key must be 32 bytes.');
    }
    if (iv.length != 16) {
      throw ArgumentError.value(iv.length, 'iv.length', 'AES-CBC IV must be 16 bytes.');
    }
  }
}

File: php/src/Crypto/OtpCipher.php
<?php

declare(strict_types=1);

namespace OtpCrypto\Crypto;

/**
 * OtpCipher – AES-256-CBC with PKCS#7 padding (OpenSSL)
 * -----------------------------------------------------
 * Thin wrapper around PHP OpenSSL to encrypt/decrypt using:
 *   - Algorithm : AES-256-CBC
 *   - Padding   : PKCS#7 (OpenSSL default, i.e., NOT using ZERO_PADDING)
 *   - Inputs    : 32-byte key (encKey) and 16-byte IV
 *
 * This class does **not** derive keys or IVs; higher layers (HKDF + IvDeriver)
 * supply them. It also does **not** compute/verify MACs; always perform
 * Encrypt-then-MAC at a higher layer (TagDeriver + constant-time compare).
 *
 * SECURITY NOTES:
 * - Enforce key length (32) and IV length (16) before calling OpenSSL.
 * - On any failure, return generic error messages to avoid leaking internals.
 * - Decrypt **only after** MAC verification has succeeded.
 */
final class OtpCipher
{
    private const CIPHER = 'aes-256-cbc';

    private function __construct() {}

    /**
     * Encrypts $plaintext with AES-256-CBC + PKCS#7.
     *
     * @param string $encKey   32-byte key (binary).
     * @param string $iv       16-byte IV (binary).
     * @param string $plaintext Binary plaintext.
     * @return string          Binary ciphertext.
     *
     * @throws \InvalidArgumentException for invalid inputs.
     * @throws \RuntimeException on encryption failure (generic message).
     */
    public static function encrypt(string $encKey, string $iv, string $plaintext): string
    {
        self::requireKeyIv($encKey, $iv);

        // OPENSSL_RAW_DATA -> return raw binary, keep PKCS#7 padding enabled.
        $ciphertext = \openssl_encrypt(
            $plaintext,
            self::CIPHER,
            $encKey,
            OPENSSL_RAW_DATA,
            $iv
        );

        if ($ciphertext === false) {
            // Do not leak OpenSSL details.
            throw new \RuntimeException('Internal error');
        }
        if ($ciphertext === '') {
            // Extremely unlikely, but reject empty output.
            throw new \RuntimeException('Internal error');
        }
        return $ciphertext;
    }

    /**
     * Decrypts $ciphertext with AES-256-CBC + PKCS#7.
     *
     * @param string $encKey     32-byte key (binary).
     * @param string $iv         16-byte IV (binary).
     * @param string $ciphertext Binary ciphertext.
     * @return string            Binary plaintext.
     *
     * @throws \InvalidArgumentException for invalid inputs.
     * @throws \RuntimeException on decryption failure (generic message).
     *
     * IMPORTANT:
     * - Callers must verify HMAC *before* calling this method.
     */
    public static function decrypt(string $encKey, string $iv, string $ciphertext): string
    {
        self::requireKeyIv($encKey, $iv);

        $plaintext = \openssl_decrypt(
            $ciphertext,
            self::CIPHER,
            $encKey,
            OPENSSL_RAW_DATA,
            $iv
        );

        if ($plaintext === false) {
            // Includes bad padding, wrong key/iv, corrupted input, etc.
            throw new \RuntimeException('Decryption failed');
        }
        return $plaintext;
    }

    // ---------------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------------

    /**
     * Validates key and IV lengths for AES-256-CBC.
     *
     * @param string $encKey
     * @param string $iv
     *
     * @throws \InvalidArgumentException if lengths are incorrect.
     */
    private static function requireKeyIv(string $encKey, string $iv): void
    {
        if (\strlen($encKey) !== 32) {
            throw new \InvalidArgumentException('encKey must be 32 bytes.');
        }
        if (\strlen($iv) !== 16) {
            throw new \InvalidArgumentException('iv must be 16 bytes.');
        }
    }
}
